cpp/include/cuspatial/experimental/detail/linestring_intersection_with_duplicates.cuh

```diff
@@ -0,0 +1,297 @@
+    auto [a, b]                   = multilinestrings1.segment(idx);
+    auto const geometry_idx       = multilinestrings1.geometry_idx_from_part_idx(part_idx);
+    auto const multilinestring2   = multilinestrings2[geometry_idx];
+    auto const geometry_collection_offset =
+      num_points_offsets_first[geometry_idx] + num_segments_offsets_first[geometry_idx];
+
+    for (auto rhs_linestring_idx = 0; rhs_linestring_idx < multilinestring2.size();
+         ++rhs_linestring_idx) {
+      auto const linestring2 = multilinestring2[rhs_linestring_idx];
+      for (auto rhs_segment_idx = 0; rhs_segment_idx < linestring2.num_segments();
+           ++rhs_segment_idx) {
+        auto [c, d]                   = linestring2.segment(rhs_segment_idx);
+        auto [point_opt, segment_opt] = segment_intersection(segment<T>{a, b}, segment<T>{c, d});
+
+        // Writes geometry and origin IDs to output. Note that for each pair, intersecting
+        // points always precedes overlapping segments (arbitrarily).
+        if (point_opt.has_value()) {
+          auto r              = cuda::atomic_ref<count_t>{n_points_stored[geometry_idx]};
+          auto next_point_idx = r.fetch_add(1);
+          points_first[num_points_offsets_first[geometry_idx] + next_point_idx] = point_opt.value();
+          auto union_column_idx = geometry_collection_offset + next_point_idx;
+          auto point_column_idx = num_points_offsets_first[geometry_idx] + next_point_idx;
+          point_ids_range.set(point_column_idx,
+                              lhs_linestring_idx,
+                              lhs_segment_idx,
+                              rhs_linestring_idx,
+                              rhs_segment_idx);
+        } else if (segment_opt.has_value()) {
+          auto r                = cuda::atomic_ref<count_t>{n_segments_stored[geometry_idx]};
+          auto next_segment_idx = r.fetch_add(1);
+          segments_first[num_segments_offsets_first[geometry_idx] + next_segment_idx] =
+            segment_opt.value();
+          auto union_column_idx =
+            geometry_collection_offset + num_points_per_pair_first[geometry_idx] + next_segment_idx;
+          auto segment_column_idx = num_segments_offsets_first[geometry_idx] + next_segment_idx;
+          segment_ids_range.set(segment_column_idx,
+                                lhs_linestring_idx,
+                                lhs_segment_idx,
+                                rhs_linestring_idx,
+                                rhs_segment_idx);
+        }
+      }
+    }
+  }
+}
+
+/**
+ * @brief Compute intersections between multilnestrings with duplicates.
+ */
+template <typename index_t,
+          typename T,
+          typename MultiLinestringRange1,
+          typename MultiLinestringRange2>
+std::pair<linestring_intersection_intermediates<vec_2d<T>, index_t>,
+          linestring_intersection_intermediates<segment<T>, index_t>>
```

Comment:
These are owning objects, so should be wrapped with `unique_ptr` to avoid double free bugs.
```suggestion
std::pair<std::unique_ptr<linestring_intersection_intermediates<vec_2d<T>, index_t>>,
          std::unique_ptr<linestring_intersection_intermediates<segment<T>, index_t>>>
```

Reply:
In fact, I think the common practice (at least in RAPIDS) is to set each of the owning member as unique pointers. This is partly because the cudf's column utilities is built around `std::unique_ptr<cudf::column>` not `cudf::column`. The effect kinda propagates all the way to all internals.

Reply:
I think a more common option here is to use `std::unique_ptr` for all members to provide flexibility on partial moves.

Reply:
Done. I wrapped all non-temporary, owning objects with `unique_ptr`.

Reply:
`Intersection_intermediate` itself isn't wrapped with unique_ptr. But all of its member variables are.